### DISCUSSION TITLE
[Dynamics/ActuallyExplicitEuler/LUMPED] is much slower than Abaqus due to computeJacobianTag and computeResidual?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#28105] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi guys,

I know that MOOSE explicit isnt optimised, but is there a way to make `computeJacobianTag` and `computeResidual` faster for `lumped` case? where it is just solved by direct inversion

I looked into `ActuallyExplicitEuler.C` and `ExplicitTimeIntegrator.C`

`ActuallyExplicitEuler::solve()` prepares matrix into solver for all cases and feed to `ExplicitTimeIntegrator::performExplicitSolve()` (CONSISTENT/LUMPED/LUMP_PRECONDITIONED) via `computeJacobianTag` and `computeResidual`

```
void
ActuallyExplicitEuler::solve()
{

  // Reset iteration counts
  _n_nonlinear_iterations = 0;
  _n_linear_iterations = 0;

  _current_time = _fe_problem.time();

  // Set time to the time at which to evaluate the residual
  _fe_problem.time() = _fe_problem.timeOld();
  _nonlinear_implicit_system->update();

  // Compute the residual
  _explicit_residual.zero();
  _fe_problem.computeResidual(
      *_nonlinear_implicit_system->current_local_solution, _explicit_residual, _nl.number());

  // Move the residual to the RHS
  _explicit_residual *= -1.0;

  // Compute the mass matrix
  auto & mass_matrix = _nonlinear_implicit_system->get_system_matrix();
  if (!_constant_mass || (_constant_mass && _t_step == 1))
    _fe_problem.computeJacobianTag(
        *_nonlinear_implicit_system->current_local_solution, mass_matrix, _Ke_time_tag);

  // Perform the linear solve
  bool converged = performExplicitSolve(mass_matrix);
```

I just wondering, is it necessary for LUMPED? Note that, i checked `ExplicitTimeIntegrator::performExplicitSolve()` , it's very fast, i dont think this is a problem at all.

At the moment, from our comparison, MOOSE explicit is 10x slower than Abaqus with the same number of cores for exactly the same case (LUMPED), I think if we make `computeJacobianTag` and `computeResidual` more efficient, MOOSE will be much faster.

below are all the kernels i am currently using (which will be more complicated in the future), it should be fast and simple

```
  [Kernels]
    [./DynamicSolidMechanics] # zeta*K*vel + K * disp
      displacements = 'disp_x disp_y disp_z'
      use_displaced_mesh=false
      eigenstrain_names = ini_stress
      strain = SMALL
      incremental = true
      # decomposition_method = HughesWinget
    [../]

    [./inertia_x]
      type = InertialForce
      use_displaced_mesh=false
      variable = disp_x
      density_scaling = density_scaling
    [../]
    [./inertia_y]
      type = InertialForce
      use_displaced_mesh=false
      variable = disp_y
      density_scaling = density_scaling
    [../]
    [./inertia_z]
      type = InertialForce
      use_displaced_mesh=false
      variable = disp_z
      density_scaling = density_scaling
    [../]

    [./gravity]
        type = Gravity
        use_displaced_mesh=false
        variable = disp_z
        function = '-9.81*if(t=0, 0, if(t>=30, 1, (t) / 30))'
        value = 1
    [../]
  []
```

Thank you!

Kind regards,
Traiwit

**Comment by GiudGiud** [comment]
Hello

Please follow these instructions to get us a profile of your simulation
https://mooseframework.inl.gov/application_development/profiling.html

From there can see if there is a hot spot we should look to improve
Guillaume

**Reply by cbolisetti** [reply]
Hey @Traiwit sorry I haven't responded in the previous discussion. If performance is the main concern, I think Derek Gaston looked at this very comparison a while ago (don't think he went too deep) and AFAIK the conclusion was that there were several little optimizations required to compete with ABAQUS or LS-DYNA. Note that these commercial codes are extremely optimized for this problem. Whereas MOOSE wasn't originally envisioned for even implicit dynamics; explicit is even newer. Given this I don't think it's a quick solution to increase the speed. It might take a consistent effort to optimize at multiple levels in the system. @recuero @dschwen are you guys planning to improve performance of the explicit solver in any of your projects?

**Reply by Traiwit** [reply]
Hi @cbolisetti,

I feel like for `ActuallyExplicitEuler` + `LUMPED` case we are close as we don't even need to use solver. we probs just need to change the way that we build matrix for this particular case before feeding into `performExplicitSolve(mass_matrix)`.

I can start looking into this, but not exactly sure where.

I see @friedmud and @hugary1995 have also involved with Explicit quite a fair bit as well.

Thank you guys! we got this :)

**Comment by Traiwit** [comment]
Hi @GiudGiud 

i'm having hard time following the instruction there, first stupid question: do I do all that profiling inside conda moose environment or outside?

at the moment, i cannot run 

`scripts/update_and_rebuild_petsc.sh --CFLAGS=-fno-omit-frame-pointer --CXX_CXXFLAGS=-fno-omit-frame-pointer` outside conda

and I cannot run `scripts/update_and_rebuild_libmesh.sh` inside conda

**Reply by GiudGiud** [reply]
for now let's just profile moose so start from inside a conda environment, and build the oprof version of moose

if needed you can try to profile the whole stack. in that case, why cannot you run this outside of conda?

> scripts/update_and_rebuild_petsc.sh --CFLAGS=-fno-omit-frame-pointer --CXX_CXXFLAGS=-fno-omit-frame-pointer outside conda

**Reply by Traiwit** [reply]
Hi @GiudGiud 

i'm pushing it through this part now

```
export GPERF_DIR=$HOME/gperftools/installed
export METHOD=oprof
cd [your-moose-app-repository]
make -j$MOOSE_JOBS
```

and got this error

```
Linking Library /home/moose/project/moose/framework/contrib/pcre/libpcre-oprof.la...
Linking Library /home/moose/project/moose/framework/contrib/gtest/libgtest.la...
/usr/bin/ld: cannot find -lmesh_oprof: No such file or directory
/usr/bin/ld: cannot find -ltimpi_oprof: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [/home/moose/project/moose/framework/moose.mk:404: /home/moose/project/moose/framework/contrib/pcre/libpcre-oprof.la] Error 1
make: *** Waiting for unfinished jobs....
make[1]: Entering directory '/home/moose/project/moose/framework/contrib/hit'
Building hit for python with python3-config
make[1]: 'hit' is up to date.
make[1]: Leaving directory '/home/moose/project/moose/framework/contrib/hit'
(base) moose@moose-AS-5014A-TT:~/project/mech_be$ 
```

I will try again on Monday, otherwise, my input file is very simple, if you have time could you please do the profiling for me?

at the moment 1 timestep take ~5-10 sec for full scale problem, but the problem is that we need to run more than 100k steps for each run, and this will take days 

```
# mpiexec -n 4 ./mech_be-opt -i ./mech_be/BE_EXAMPLE/central_diff_massscaling.i

# Test for the central difference time integrator in 3D.
[GlobalParams]
use_displaced_mesh=false
[]

[Mesh]
  parallel_type = distributed
[gen]
  type = GeneratedMeshGenerator
  dim = 3 # Dimension of the mesh
  nx = 10 # Number of elements in the x direction
  ny = 10 # Number of elements in the y direction
  nz = 20 # Number of elements in the z direction
  xmin = 0.0
  xmax = 1000
  ymin = 0.0
  ymax = 1000
  zmin = 0.0
  zmax = 1000
  # allow_renumbering = false # So NodalVariableValue can index by id
  show_info = true
  # elem_type = HEX8
  elem_type = TET10
[]
[]

  
  [Variables]
    [./disp_x]
    [../]
    [./disp_y]
    [../]
    [./disp_z]
    [../]
  []
  
  [AuxVariables]
    [./vel_x]
    [../]
    [./accel_x]
    [../]
    [./vel_y]
    [../]
    [./accel_y]
    [../]
    [./vel_z]
    [../]
    [./accel_z]
    [../]
  []
  
  [AuxKernels]
    [./accel_x]
      type = TestNewmarkTI
      variable = accel_x
      displacement = disp_x
      first = true
    [../]
    [./vel_x]
      type = TestNewmarkTI
      variable = vel_x
      displacement = disp_x
      first = true
    [../]

    [./accel_y]
      type = TestNewmarkTI
      variable = accel_y
      displacement = disp_y
      first = true
    [../]
    [./vel_y]
      type = TestNewmarkTI
      variable = vel_y
      displacement = disp_y
      first = true
    [../]

    [./accel_z]
      type = TestNewmarkTI
      variable = accel_z
      displacement = disp_z
      first = true
    [../]
    [./vel_z]
      type = TestNewmarkTI
      variable = vel_z
      displacement = disp_z
      first = true
    [../]
  []

  [Kernels]
    [./DynamicSolidMechanics] # zeta*K*vel + K * disp
      displacements = 'disp_x disp_y disp_z'
      use_displaced_mesh=false
      eigenstrain_names = ini_stress
      strain = SMALL
      incremental = true
      # decomposition_method = HughesWinget
    [../]

    [./inertia_x]
      type = InertialForce
      use_displaced_mesh=false
      variable = disp_x
      density_scaling = density_scaling
    [../]
    [./inertia_y]
      type = InertialForce
      use_displaced_mesh=false
      variable = disp_y
      density_scaling = density_scaling
    [../]
    [./inertia_z]
      type = InertialForce
      use_displaced_mesh=false
      variable = disp_z
      density_scaling = density_scaling
    [../]

    [./gravity]
        type = Gravity
        use_displaced_mesh=false
        variable = disp_z
        function = '-9.81*if(t=0, 0, if(t>=30, 1, (t) / 30))'
        value = 1
    [../]
  []


  
  [BCs]

    [./no_x_back]
      type = DirichletBC
          preset = false
      variable = disp_x
      boundary = 'back'
      value = 0.0
    [../]
    [./no_y_back]
      type = DirichletBC
          preset = false
      variable = disp_y
      boundary = 'back'
      value = 0.0
    [../]
    [./no_z_back]
      type = DirichletBC
          preset = false
      variable = disp_z
      boundary = 'back'
      value = 0.0
    [../]

   # fixed x left
    [./no_x_left]
      type = DirichletBC
          preset = false
      variable = disp_x
      boundary = 'left'
      value = 0.0
    [../]
    [./no_y_left]
      type = DirichletBC
          preset = false
      variable = disp_y
      boundary = 'left'
      value = 0.0
    [../]
      

    # fixed x right
     [./no_x_right]
       type = DirichletBC
          preset = false
       variable = disp_x
       boundary = 'right'
       value = 0.0
     [../]
     [./no_y_right]
       type = DirichletBC
          preset = false
       variable = disp_y
       boundary = 'right'
       value = 0.0
     [../]

     # fixed y top
      [./no_x_top]
        type = DirichletBC
          preset = false
        variable = disp_x
        boundary = 'top'
        value = 0.0
      [../]
      [./no_y_top]
        type = DirichletBC
          preset = false
        variable = disp_y
        boundary = 'top'
        value = 0.0
      [../]

      # fixed y bottom
       [./no_x_bottom]
         type = DirichletBC
          preset = false
         variable = disp_x
         boundary = 'bottom'
         value = 0.0
       [../]
       [./no_y_bottom]
         type = DirichletBC
          preset = false
         variable = disp_y
         boundary = 'bottom'
         value = 0.0
       [../]
  []

  
   # * if(t=0, 0, if(t>=(target_time), 1, (t) / (target_time) ))
  [Functions]
    [./ini_xx]
        type = ParsedFunction
        expression = '-1.5*9.81*2700*(1000-z)*if(t=0, 0, if(t>=30, 1, (t) / 305))'
      [../]
      [./ini_yy]
        type = ParsedFunction
        expression = '-1.5*9.81*2700*(1000-z)*if(t=0, 0, if(t>=30, 1, (t) / 30))'
      [../]
      [./ini_zz]
        type = ParsedFunction
        expression = '-9.81*2700*(1000-z)*if(t=0, 0, if(t>=30, 1, (t) / 30))'
      [../]
  []
  
  [Materials]
    [./elasticity_tensor]
      youngs_modulus = 20e9 #Pa
      poissons_ratio = 0.25
      type = ComputeIsotropicElasticityTensor
    [../]
    [./strain]
    #   type = ComputeFiniteStrain
      type = ComputeIncrementalStrain
      eigenstrain_names = ini_stress
      displacements = 'disp_x disp_y disp_z'
      implicit = false
    [../]
    [./stress]
      type = ComputeFiniteStrainElasticStress
    #   type = ComputeLinearElasticStress
    [../]
    [./density]
      type = GenericConstantMaterial
      prop_names = density
      prop_values = 2700 #kg/m3
    [../]
    [./ini_stress]
      type = ComputeEigenstrainFromInitialStress
      eigenstrain_name = ini_stress
      initial_stress = 'ini_xx 0 0 0 ini_yy 0 0 0 ini_zz'
    [../]
    [density_scaling]
      type = DensityScaling
      density = density
      desired_time_step = 0.0005
      output_properties = density_scaling
      # outputs = 'exodus'
      factor = 0.2
    []
  []
  
  [Executioner]
    type = Transient
    start_time = 0
    end_time = 40
    num_steps = 20
    dt = 0.0005
    [./TimeIntegrator]
      type = CentralDifference
      use_constant_mass = true
      solve_type = lumped
    [../]
    # [TimeStepper]
    #   type = PostprocessorDT
    #   postprocessor = time_step
    # []
  []
  
  
  [Outputs]
    # exodus = true
  []

```

**Reply by GiudGiud** [reply]
> /usr/bin/ld: cannot find -lmesh_oprof: No such file or directory
/usr/bin/ld: cannot find -ltimpi_oprof: No such file or directory

ok the conda path is not going to work. I thought we were shipping oprof libraries too.
Maybe try `conda info moose` to see if there is a 'prof' named package

**Reply by GiudGiud** [reply]
I won't have time to look this weekend, it looks like building from source outside the conda environment, or inside a reduced conda environment with just moose-mpich is going to be the way to build this

**Comment by Traiwit** [comment]
Hi guys, a quick update:
 
I tried to optimise MOOSE source code today (mostly all cleaned up the kernels)
 
a mesh with 5,873,043 nodes and 4,361,842 elements 2nd order,  40s (dt = 0.005) took ~ 24 hrs before now down to ~15 hrs (this is for 32 cores pc)
 
I still think the major issue is where they put together all the kernels and build the matrix via `computeResidual()` - I still couldnt understand this part of their code, very complicated

I tried to have all the kernels off (commented all of them out and run with `kernel_coverage_check = false`), the simulation run super fast, and when i tried to have a fake kernel, for example

```
GravityTempl<is_ad>::computeQpResidual()
{
  return 0;
}
```

the simulation become as slow as before, so i guess the problem is pretty much come down to the way MOOSE put together the kernels and build the matrix

**Reply by hugary1995** [reply]
I think your intuition is correct. Here's my suggestion as I think you are basically doing profiling the hard way, i.e., by recording wall time from huge runs with different configurations. Here is the easy way, as @GiudGiud suggested at the beginning:
1. Compile your MOOSE with gperftools following the instructions.
2. Run a small-ish problem and with gperftools cpuprofiler. Only a few thousand dofs is fine, just make sure the total wall time is at least a few seconds (the longer the better, apparently, as the cpuprofiler is a sampling based profiler).
3. Use pprof to list hot spots, from where we can identify code regions that could be optimized.

**Reply by hugary1995** [reply]
Regardless, the last time I looked at this, I noticed a few ways of speeding things up at the framework level:
1. Matrix contribution from linear kernels can be entirely skipped. We just need to compute them once at the beginning of the entire simulation. These kernels include inertial force, gravity, linear damping, and even the stress-divergence kernel in the case of linear elasticity. This will result in huge savings. Implementing this would require the tagging system, which wasn't mature at the time we looked at this issue, but this should be doable nowadays.
2. MOOSE materials are evaluated on-the-fly -- meaning that they are computed along with the element loop. Ideally, we only need to reinit materials that are consumed by the kernels (together with the first bullet point, we can skip many material evaluations), but currently we always reinit all materials. This wasn't a big issue for implicit (which is what MOOSE was originally designed for), but this dominates explicit dynamics wall time.
3. The auxkernels for accelerations and velocities only need to be executed at LINEAR.

IIRC, we have a `constant_mass` option which is basically a specialized case for the first bullet point.

**Reply by hugary1995** [reply]
I'd also like to add that resolving the second bullet point also benefits MOOSE implicit.

**Reply by Traiwit** [reply]
Hi @hugary1995 

Thank you for you suggestions! they are very helpful.

I will try to do the profiling again, then I will look into the 2nd bullet point, and the 3rd point should be as simple as just fixing the input file I assume.

I will keep you guys updated :)

**Reply by GiudGiud** [reply]
1 should be the easiest to do.

For porous flow you have quite a few linear terms right?
BUT as you mentioned adding a single term already slows it down too much so 1) is not enough for your case UNLESS all your kernels are linear

**Reply by Traiwit** [reply]
Hi guys

I've been working on 2) now the run time is ~30% faster

I have a few questions to ask:

@GiudGiud 

what is `FEProblem::outputStep` why does it take so long?

even i turned off the whole [Outputs] block, it's still there, is this related to data exchange after the matrix is solved?

![image](https://github.com/user-attachments/assets/b5099991-6eeb-4bc4-95a6-58cf185d2c21)
```
void
FEProblemBase::outputStep(ExecFlagType type)
{
  TIME_SECTION("outputStep", 1, "Outputting");

  setCurrentExecuteOnFlag(type);

  for (auto & sys : _solver_systems)
    sys->update();
  _aux->update();
  if (_displaced_problem)
    _displaced_problem->syncSolutions();
  _app.getOutputWarehouse().outputStep(type);

  setCurrentExecuteOnFlag(EXEC_NONE);
}
```



@hugary1995 @cbolisetti I have a question regarding `constant_mass`, I checked, so when we have that equal to true, we only `computeQpJacobian()` once at `_t_step ==1`

for example below is `InertialForceTempl<false>::computeQpJacobian()`, it has `_du_dotdot_du` and `_du_dot_du` terms which are updated at every timestep (right?), i'm not sure if this is the right way to do it, shouldnt it be updated at every timestep, or I missed something important here? I remember @recuero mentioned this before, but I didn't understand it back then.

```
      return _test[_i][_qp] * _density[_qp] * (*_du_dotdot_du)[_qp] * _phi[this->_j][_qp] +
             _eta[_qp] * (1 + _alpha) * _test[_i][_qp] * _density[_qp] * (*_du_dot_du)[_qp] *
                 _phi[this->_j][_qp];
```

Thank you guys!

**Reply by GiudGiud** [reply]
Even when you turn everything off there is still the automated checkpoint
set wall_time_checkpoint = false
https://mooseframework.inl.gov/source/outputs/Checkpoint.html

Did you end up profiling with a profiler?

**Reply by Traiwit** [reply]
Hi @GiudGiud 

![image](https://github.com/user-attachments/assets/acc2ef18-10a9-43d0-84c4-c5efe48a7cd8)

above is 20 steps version, dont think it helps much, i already set that to false before I posted it. not sure why outputStep is as big as 1/4 of the solver time.

here is my [Outputs]

```
  [Outputs]
    checkpoint = false
    perf_graph = true
    wall_time_checkpoint = false
    # console = false
    # exodus = true
    # [console]
    #   type = Console
    #   output_screen = false
    # []
    [Checkpoint_out]
      type = Checkpoint
      num_files = 0
      wall_time_checkpoint = false
      enable = false
      execute_on = FINAL
    []
  []
```

RE: profiling, not yet, I can try it again (but i dont have much hope for this)

**Reply by GiudGiud** [reply]
```
void
FEProblemBase::outputStep(ExecFlagType type)
{
  TIME_SECTION("outputStep", 1, "Outputting");

  setCurrentExecuteOnFlag(type);

  for (auto & sys : _solver_systems)
    sys->update();
  _aux->update();
  if (_displaced_problem)
    _displaced_problem->syncSolutions();
```
this is probably the slow part (though a profiler will tell you for sure) (or comment them out and see if it gets out of the perf_graph)
these sync operations will cost you in parallel and when using the displaced mesh. We would need to detect whether the need is real (probably let the outputWarehouse make that determination based on what outputs have been created)

**Reply by Traiwit** [reply]
RE:  `_du_dotdot_du` and `_du_dot_du`

i just realised they are a constant from `CentralDifference::computeTimeDerivatives()`

```
 // used for Jacobian calculations
  _du_dot_du = 1.0 / (2 * _dt);
  _du_dotdot_du = 1.0 / (_dt * _dt);
```

@GiudGiud the problem is from 

`_app.getOutputWarehouse().outputStep(type);` under `FEProblemBase::outputStep(ExecFlagType type)`

yeh not sure what to do, maybe i should call it at a certain interval of time (like every 200 steps something like this) as I guess, it is important to output .e and other stuff

**Reply by GiudGiud** [reply]
> @GiudGiud the problem is from

> _app.getOutputWarehouse().outputStep(type); under FEProblemBase::outputStep(ExecFlagType type)

that does not make sense if you have turned off all ooutputs before. The warehouse calls output on each outputting object
Maybe you kept exodus output?

**Reply by YaqiWang** [reply]
For the second bullet, you can split the code in your materials into user objects, which have the control via execute_on. These user objects can be built internally by the material and do the heavy-lifting evaluations.
